### ISSUE TITLE
What is D_noise for?

### ISSUE THREAD

**Issue by rainsoulsrx** [oneThousand1000/HairMapper#3]
Thank you for your excellent work.
I wonder what the D_0 and D_noise for, why you generate two part of random images?

**Comment by oneThousand1000** [comment]
Hi rainsoulsrx, please refer to Sec 3.2 in the paper.

The e4e encoder can map a real image into its latent code w+,  w+ consists of a series of latent vectors w with low variance, each ** close to ** the distribution of the W latent space of StyleGAN. 
In order to apply our method to latent codes of real images using e4e, we sample two w+ latent code set: D0 (in W latent space) and Dnoise (add noise to codes in each layer, ** close to ** the distribution of the W latent space).

**Comment by rainsoulsrx** [comment]
Thanks for you quick reply.
You mean you train on both D0 and Dnoise separately, in order to make the model robust in both stylegan fake images and real images?

**Comment by rainsoulsrx** [comment]
And I run the code, and found some failures like follows:
<img width="1307" alt="image" src="https://user-images.githubusercontent.com/24955512/165314074-9f4f4daa-ff76-488f-a14c-4d5f9e6f3a05.png">
Is this right?

**Comment by oneThousand1000** [comment]
> Thanks for you quick reply.
> You mean you train on both D0 and Dnoise separately, in order to make the model robust in both stylegan fake images and real images?

Yes, you are right.

**Comment by oneThousand1000** [comment]
> And I run the code, and found some failures like follows: <img alt="image" width="1307" src="https://user-images.githubusercontent.com/24955512/165314074-9f4f4daa-ff76-488f-a14c-4d5f9e6f3a05.png"> Is this right?

The failures are acceptable. During training, some images' hair can not be removed completely (about 5%) by the hair separation boundary, that's also the reason why we use the paired latent codes to train a hair mapper (the failures won't affect the training of hair mapper).

**Comment by oneThousand1000** [comment]
There are also some images in my own training dataset that still have hair, they didn't affect the training of the hair mapper.
![img](https://user-images.githubusercontent.com/32099648/165316570-bdbdb24f-27a0-46ef-a772-978588f41ab1.PNG)

**Comment by rainsoulsrx** [comment]
> There are also some images in my own training dataset that still have hair, they didn't affect the training of the hair mapper. ![img](https://user-images.githubusercontent.com/32099648/165316570-bdbdb24f-27a0-46ef-a772-978588f41ab1.PNG)

Got it!! Thank you for your kind and detailed reply~~

**Comment by rainsoulsrx** [comment]
Hi, I have another question.
After you get the edited image, you optimize the result using L1 and VGG loss, when using L1 loss, you multi hair mask, I think it is reasonable. But when calculating VGG loss, you do not multi the mask, why?
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/24955512/165320522-7fbba5a5-a845-4b74-b90b-e2c09dc6bf1f.png">

**Comment by oneThousand1000** [comment]
> Hi, I have another question. After you get the edited image, you optimize the result using L1 and VGG loss, when using L1 loss, you multi hair mask, I think it is reasonable. But when calculating VGG loss, you do not multi the mask, why? <img alt="image" width="1008" src="https://user-images.githubusercontent.com/24955512/165320522-7fbba5a5-a845-4b74-b90b-e2c09dc6bf1f.png">

We use vgg loss (perceptual loss) to penalize the high-level (geometric) feature difference between x and x_rec, we want the x_rec to maintain the geometric feature of the bald head and face in x (the edited image).
Please refer to Sec 3.4 in [Coarse-to-Fine: Facial Structure Editing of Portrait Images via Latent Space Classifications](http://www.cad.zju.edu.cn/home/jin/sig2021/paper46.pdf), we explained the details of diffusion.

**Comment by rainsoulsrx** [comment]
> 

Oh oh oh, I git it, thanks!!!
